Return a non-zero exit code for all unhandled errors

JayKickliter opened the pull request:
`main()` returns a result only for errors we can't easily intercept and log. An example is config file parsing. The config file is the source of truth for the kind of logger we need to build. Any `Err`'s returned are displayed to STDERR and result in a non-zero exit code. And due to the behavior of our logger, simply calling `exit()` early prevents any error logging from reaching its destination.

The solution is an immediately invoked lambda expression, which is the program's exit code.

An alternative _could_ be to have yet another intermediate 'runner' fn, but I felt that would be just as confusing and messy as it would result in this: `main() → new_runner() → run()`.

xandkar reviewed (approved):
Is the intermediate closure strictly necessary? If so, then why? Would not `let retcode = {...};` be sufficient?

JayKickliter commented:
@xandkar

> Is the intermediate closure strictly necessary? If so, then why? 

It _was_ necessary at first, when I was handling runtime build failure. I switched that to `expect()` because it's such an uncommon error, and nothing can be done when it happens. 

> Would not let retcode = {...}; be sufficient?

Yep! I updated the code to that. Thanks.

JayKickliter commented on src/main.rs:
```diff
@@ -87,37 +87,58 @@ pub fn main() -> Result {
+    let retcode = {
+        || {
+            let logger = mk_logger(&settings);
+            let _scope_guard = slog_scope::set_global_logger(logger);
```
Yep! The global logger will only live as long as the guard exists. You'd see this `_<THING>_guard` pattern in Rust & C++. So the global value it guards is used, even if the bound value isn't used directly. 

FYI: rust made the questionable decision that binding to `_` causes the value to immediately call `Drop::drop()`, and is a common source of bugs for new users.

jeffgrunewald replied:
👍

xandkar replied:
> Yep! The global logger will only live as long as the guard exists. So the global value it guards is used, even if the bound value isn't used directly. You'll see this `_<THING>_guard` pattern in Rust & C++.
> 
> FYI: rust made the questionable decision that binding to `_` causes the value to immediately call `Drop::drop()`, and is a common source of bugs for new users.

Holy crap - this is _really_ non-obvious! Great digging!

xandkar replied:
@JayKickliter an `XXX` comment with this reason would probably be appreciated by future readers.

JayKickliter replied:
Yep. I took it for granted that this pattern was commonly understood, but of course it isn't

xandkar replied:
I was mad at Erlang for not treating `_` and `_foo` exactly the same, but this is way worse.

xandkar replied:
Looks like something different than what I initially thought is happening. It seems the main issue is reference counting. The global logger is a `static`, so should _not_ be auto-dropped. However, that `static` is _not_ the _logger_ directly, but is a container - `Arc`, which itself contains the logger and will drop it on absence of current references (which are decremented on `let _ = ...;`).

@JayKickliter does sound right or am I still off on what's happening here?

JayKickliter replied:
Your analysis could accurate. I haven't looked at the implementation.
